text/3512-closure-move-bindings.md

```diff
@@ -0,0 +1,284 @@
+    move || run(foo, bar)
+}
+```
+
+- It is a very frequent scenario to clone a value into a closure
+(especially common with `Rc`/`Arc`-based values),
+but even the simplest scenario requires three lines of boilerplate:
+
+```rs
+{
+    let foo = foo.clone();
+    move || foo.run()
+}
+```
+
+This RFC proposes a more concise syntax to express these moving semantics.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A closure may capture bindings in its defining scope.
+Bindings are captured by reference by default:
+
+```rs
+let mut foo = 1;
+let mut closure = || { foo = 2; };
+closure();
+dbg!(foo); // foo is now 2
+```
+
+You can add a `move` keyword in front of the closure
+to indicate that all captured bindings are moved into the closure
+instead of referenced:
+
+```rs
+let mut foo = 1;
+let mut closure = move || { foo = 2; };
+closure();
+dbg!(foo); // foo is still 1, but the copy of `foo` in `closure` is 2
+```
+
+Note that `foo` is _copied_ during move in this example
+as `i32` implements `Copy`.
+
+If a closure captures multiple bindings,
+the `move` keyword makes them all captured by moving.
+To move only specific bindings,
+list them in parentheses after `move`:
+
+```rs
+let foo = 1;
+let mut bar = 2;
+let mut closure = move(mut foo) || {
+    foo += 10;
+    bar += 10;
+};
+closure();
+dbg!(foo, bar); // foo = 1, bar = 12
+```
+
+Note that the outer `foo` no longer requires `mut`;
+it is relocated to the closure since it defines a new binding.
+
+Moved bindings may also be renamed:
+
+```rs
+let mut foo = 1;
+let mut closure = move(mut bar = foo) || {
+    foo = 2;
+    bar = 3;
+};
+closure();
+dbg!(foo); // the outer `foo` is 2 as it was captured by reference
+```
+
+Bindings may be transformed when moved:
+
+```rs
+let foo = vec![1];
+let mut closure = move(mut foo = foo.clone()) || {
+    foo.push(2);
+};
+closure();
+dbg!(foo); // the outer `foo` is still [1] because only the cloned copy was mutated
+```
+
+The above may be simplified to `move(mut foo.clone())` as well.
+This simplification is only allowed
+when the transformation expression is a method call on the captured binding.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A closure expression has the following syntax:
+
+> **<sup>Syntax</sup>**\
+> _ClosureExpression_ :\
+> &nbsp;&nbsp; ( `move` _MoveBindings_<sup>?</sup> )<sup>?</sup>\
+> &nbsp;&nbsp; ( `||` | `|` _ClosureParameters_<sup>?</sup> `|`  )\
+> &nbsp;&nbsp; (_Expression_ | `->` _TypeNoBounds_&nbsp;_BlockExpression_)>\
+> _MoveBindings_ :\
+> &nbsp;&nbsp; `(` ( _MoveBinding_ (`,` _MoveBinding_)<sup>\*</sup> `,`<sup>?</sup> )<sup>?</sup> `)`\
+> _MoveBinding_ :\
+> &nbsp;&nbsp; _NamedMoveBinding_ | _UnnamedMoveBinding_\
+> _NamedMoveBinding_ :\
+> &nbsp;&nbsp; _PatternNoTopAlt_ `=` _Expression_\
+> _UnnamedMoveBinding_ :\
+> &nbsp;&nbsp; `mut`<sup>?</sup> ( _IdentifierExpression_ | _MethodCallExpression_ )\
```

Comment:
I think we should also have `move(&x, &mut y)` syntax, for the specific cases of explicit by-reference captures:

  - not only for my suggested implicit-by-`move` change, should it be accepted;
  - but even in the case of the current implict-by-ref-captures semantics of the RFC, when wanting to be explicit, it would be nicer to be using `move(&x, &mut y)` than `move(x = &x, y = &mut y)`, especially when `x` and `y` are `longer_variable_names`.
  - there would otherwise be no way to mimic, explicitly, the "magic binding" semantics which implicit by-ref closures captures currently have (_e.g._, when having `x: i32`, an implicit by-ref capture of `x` makes it so the closure has captured `&x`, but the `x` name is still fully usable to refer to the `*&x` place).

If we didn't have this sugar, I could very well see myself having a `.by_ref()` and `.by_mut()` couple of blanket-implemented extension methods.
  - whilst these methods could be a bearable alternative, I don't think it's worth the hassle w.r.t. adding another 1-`lookahead` case for these parsing rules:

```rs
    $binding:ident $(. $method…)?
OR
    &$(mut)? $binding:ident
OR
    $binding:pat = $e:expr
```